### ISSUE TITLE
kin cluster stalled after seeing WouldExceedMaxAccountCostLimit error

### ISSUE THREAD

**Issue by HaoranYi** [solana-labs/solana#31340]
#### Problem

Recently, we run into a network stall on kin-clusters.

After seeing the following errors, the cluster stalled. 

```
[2023-04-21T14:21:18.360233163Z ERROR solana_metrics::metrics] datapoint: replay-stage-mark_dead_slot error="error: InvalidTransaction(WouldExceedMaxAccountCostLimit)" slot=249053i
```

```
[2023-04-21T14:21:34.745678677Z INFO  solana_core::progress_map] Propagated stats:
                    total staked: 3999990868480,
                    observed staked: 999997717120,
                    vote pubkeys: {2TLD7hwoUvxS3ufmUNMsBvfaL8Q3XjhYmbvHJCoEuBke},
                    node_pubkeys: {31fxZovs3gBKVVTtC2VJUuKeVoq6mQkLjWnicWhErQ4f},
                    slot: 249067,
                    epoch: Some(12)
```

#### Proposed Solution

Debug and fix

**Comment by taozhu-chicago** [comment]
Banking stage executes and pack transactions separately from replay stage. But both use `requested units` from transactions, so the accumulated CUs should be stable. Sounds like there are edge cases. I'll run test cluster on gce, see if I can reproduce it. But if anyone have better idea to repro/debug, please let me know.

**Comment by CriesofCarrots** [comment]
> But if anyone have better idea to repro/debug, please let me know.

Have we already used ledger-tool examined that slot on the leader vs another node, and compared hashes, capitalization, etc? Seems like an easy place to start.

**Comment by taozhu-chicago** [comment]
Taking slot `249559` as example, 

It is produced by `kin1`, log indicates slot 249559 frozen:
```
[2023-04-21T18:40:24.178204033Z INFO  solana_runtime::bank] bank frozen: 249559 hash: CyAtbbiiUsrv186uHk28ywAmAkQSygzpuiXozVr1kTjC accounts_delta: 7NttyhJoFmWFy3TezxwWEDhbkEABUpTFTxNh4WnY5sBj signature_count: 269 last_blockhash: 3zErKhExj5mtdgQTDKUCXj9iFExnjf9bNgXG11VqPPRV capitalization: 1000126221835756359, stats: BankHashStats ...
```

On `kin2` (this is also the same on kin3,4  just with slightly different timestamps), replaying slot 249559 failed due to account cost limit: 
```
[2023-04-21T18:40:24.058107838Z ERROR solana_metrics::metrics] datapoint: replay-stage-mark_dead_slot error="error: InvalidTransaction(WouldExceedMaxAccountCostLimit)" slot=249559i
```

Using ledger-tool to replay,
from `kin1`, replay would fail for same cause as kin2 in real time. 
```
[2023-04-25T18:19:31.819582676Z WARN  solana_ledger::blockstore_processor] slot 249559 failed to verify: invalid transaction error: Transaction would exceed max account limit within the block
```

from `kin2`:
```
[2023-04-25T17:55:16.545572899Z WARN  solana_ledger::blockstore_processor] slot 249559 failed to verify: failed to load entries, error: blockstore error
```

All this is to say the inconsistency is between banking stage and replay stage.

**Comment by taozhu-chicago** [comment]
TLDR: the feature in replay (https://github.com/solana-labs/solana/blob/master/ledger/src/blockstore_processor.rs#L351-L361) cannot be enabled, yet.

The [QoS adjustment logic](https://github.com/solana-labs/solana/blob/master/core/src/banking_stage/consumer.rs#L427-L432) still alive in leader code. It should have been removed after cost model replaced "estimated cost" with "requested cu". It was kept because user can send TX with Max 1.4M CUs without additional cost, such TXs would degrade cluster. 

A possible path forward is: 
Step 1: replace prioritization from `micro-lamports per CU` to `lamports per CU`, amount other reasons, to increase cost of additional CUs in TX;
Step 2: furnishing `fee_structure`, so transaction base fee reflects its total CUs;
Step 3: with above economical means, QoS Adjustment can be removed;

PS, Thanks @HaoranYi for helping debugging

**Comment by 7layermagik** [comment]
For step 4, do you guys do any longer term correlation type of experiments between cu usage and replay times? I remember confirmation times increasing as avg block cu usage went up on mainnetbeta -- I suspect the replay times would correlate moreso to certain hotspots hitting the account cu limit. Adding more threads-worth of blockspace will be good in preparation for later.... so we don't get any high-fee shocks if there's a ton of surprise adoption.

Anyway, looking forward to these changes :) I'm sure there will be some debate around what the base fee per cu should be for Step 2.

**Comment by taozhu-chicago** [comment]
We can aggregate metrics from mainnet-beta to get an cluster average of CU/micro-sec ratio, excluding fastest and slowest nodes. With that we can calculate block CU limit = block_time * CU/us * concurrency. Somewhat documented [here]( https://github.com/solana-labs/solana/blob/master/runtime/src/block_cost_limits.rs)

As for base fees, there are discussions at [SIMD 17](https://github.com/solana-foundation/solana-improvement-documents/pull/19), and [SIMD 4](https://github.com/solana-foundation/solana-improvement-documents/pull/4). Please join the debate there 😄

**Comment by KirillLykov** [comment]
While performing testing of another issue I've found this problem.
In particular, all the nodes except for one were stopped due to panic with the following error:

```
[2023-05-12T11:51:08.747642995Z ERROR solana_metrics::metrics] datapoint: replay-stage-mark_dead_slot error="error: InvalidTransaction(WouldExceedMaxBlockCostLimit)" slot=3544i
[2023-05-12T11:51:08.747653569Z INFO  solana_core::cluster_slot_state_verifier] check_slot_agrees_with_cluster() slot: 3544, root: 3512, slot_state_update: Dead(DeadState { cluster_confirmed_hash: Some(DuplicateConfirmed(JBD4GxBMYKZdxNFbhxPSZv7FvAShPBvg8dSbCdZYf1gw)), is_slot_duplicate: false })
[2023-05-12T11:51:08.747683107Z WARN  solana_core::cluster_slot_state_verifier] Cluster duplicate confirmed slot 3544 with hash JBD4GxBMYKZdxNFbhxPSZv7FvAShPBvg8dSbCdZYf1gw, but we marked slot dead
[2023-05-12T11:51:08.747690366Z INFO  solana_core::heaviest_subtree_fork_choice] marking fork starting at: (3544, 11111111111111111111111111111111) invalid candidate
thread 'solReplayStage' panicked at 'We have tried to repair duplicate slot: 3544 more than 10 times and are unable to freeze a block with bankhash JBD4GxBMYKZdxNFbhxPSZv7FvAShPBvg8dSbCdZYf1gw, instead we have a block with bankhash Some(11111111111111111111111111111111). This is most likely a bug in the runtime. At this point manual intervention is needed to make progress. Exiting', core/src/replay_stage.rs:1312:25
stack backtrace:
[2023-05-12T11:51:08.874144138Z INFO  solana_metrics::metrics] datapoint: shred_insert_is_full total_time_ms=357i slot=3560i last_index=30i num_repaired=0i num_recovered=6i
[2023-05-12T11:51:08.874176807Z INFO  solana_metrics::metrics] datapoint: poh_slot_timing slot=3560i start_time=1683892268491i end_time=1683892268836i full_time=1683892268874i start_to_full_time_diff=383i full_to_end_time_diff=-38i
   0: rust_begin_unwind
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/core/src/panicking.rs:64:14
   2: solana_core::replay_stage::ReplayStage::dump_then_repair_correct_slots::{{closure}}
   3: hashbrown::map::HashMap<K,V,S,A>::retain
             at /cargo/registry/src/index.crates.io-6f17d22bba15001f/hashbrown-0.12.3/src/map.rs:834:21
   4: std::collections::hash::map::HashMap<K,V,S>::retain
             at /rustc/84c898d65adf2f39a5a98507f1fe0ce10a2b8dbc/library/std/src/collections/hash/map.rs:691:9
   5: solana_core::replay_stage::ReplayStage::dump_then_repair_correct_slots
             at /home/sol/wrks_ext/tests_for_ale/solana-ghsa-27r9-44rr-jp8f/core/src/replay_stage.rs:1247:35
   6: solana_core::replay_stage::ReplayStage::new::{{closure}}
             at /home/sol/wrks_ext/tests_for_ale/solana-ghsa-27r9-44rr-jp8f/core/src/replay_stage.rs:969:17
```

They restart, but never catch up and the cluster is not operational (last - root distance is always growing).

### Affected versions

* fb7ba97afc - crash (logs and metrics are for this commit)
*  master ( c5905f525cdda44f490698e05d4e61fa40a4e835 ) - crash
* v1.13 -- no crash
* v1.14.7 -- no crash

### Setup

1. net script are used to create a cluster with 5+1 validators and 1 idle client
2. custom on-chain program 
3. typescript scripts for setting up accounts and execute program

With this, we execute transactions, each modifying  1 account of 16bytes (uses CPI).
Total number of accounts 64.
The client will send transactions asynchronously without checking confirmation for some duration of time.
From metrics tps looks to be ~ 2k tps. 
One transaction takes ~5k CU.
After couple of minutes cluster dies.
 
 ### Metrics dashboard
 
 https://internal-metrics.solana.com:8889/sources/1/dashboards/14?refresh=Paused&tempVars%5BDB_NAME%5D=testnet-dev-klykov&lower=now%28%29%20-%201h&zoomedLower=2023-05-12T11%3A47%3A33.855Z&zoomedUpper=2023-05-12T11%3A59%3A51.204Z

**Comment by t-nelson** [comment]
>  In particular, all the nodes except for one were stopped due to panic with the following error:

implying a leader/replay divergence. identify the diverging transaction and most likely the root cause will be obvious

> * v1.14.7 -- no crash

you mean 1.14.17?
